Add resumable downloads for llama-server model loading

ericcurtin opened the pull request:
- Implement resumable downloads in common_download_file_single function
- Add detection of partial download files (.downloadInProgress)
- Check server support for HTTP Range requests via Accept-Ranges header
- Implement HTTP Range request with "bytes=<start>-" header
- Open files in append mode when resuming vs create mode for new downloads

ericcurtin commented:
@ngxson @ggerganov PTAL

ericcurtin commented:
macOS x86 build still flakey, seems like it's random luck based on which build server you get, sometimes it has the right versions of things to compile the code.

ericcurtin commented on common/arg.cpp:
```diff
@@ -220,12 +226,25 @@ struct curl_slist_ptr {
     while (remaining_attempts > 0) {
         LOG_INF("%s: %s %s (attempt %d of %d)...\n", __func__ , method_name, url.c_str(), max_attempts - remaining_attempts + 1, max_attempts);
 
+        if (std::filesystem::exists(path_temporary)) {
```
@ngxson I suspect you could have written a lot of this. This could be a follow on PR. But is there an argument to be made that we should check things like "Last-Modified" header before each retry? For the rare case that the model got updated in between requests.

ericcurtin commented on common/arg.cpp:
```diff
@@ -220,12 +226,25 @@ struct curl_slist_ptr {
 #define CURL_RETRY_DELAY_SECONDS 2
 
-static bool curl_perform_with_retry(const std::string & url, CURL * curl, int max_attempts, int retry_delay_seconds, const char * method_name) {
+static bool curl_perform_with_retry(const std::string & url,
```
@ngxson I suspect you could have written a lot of this. This could be a follow on PR. But is there an argument to be made that we should check things like "Last-Modified" header before each retry? For the rare case that the model got updated in between requests.

ericcurtin commented on common/arg.cpp:
```diff
@@ -770,7 +815,7 @@ static std::string common_docker_get_token(const std::string & repo) {
 
 static std::string common_docker_resolve_model(const std::string & docker) {
-    // Parse ai/smollm2:135M-Q4_K_M
+    // Parse ai/smollm2:135M-Q4_0
```
The Q4_K_M version of the model seems temperamental, trying to recommend Q4_0 instead.

ngxson commented:
We currently check the ETag header for this, as most of the time (if not
all?) ETag is the hash of the remote file to be downloaded.

ericcurtin commented:
> We currently check the ETag header for this, as most of the time (if not all?) ETag is the hash of the remote file to be downloaded.

Sure, but we don't check between retries, there's a minor chance the content changes in between retries. It probably wasn't a big deal before, you pull a full file or you don't. But with resumable transfers it becomes more relevant, because you could have half one version of the file, half a different version of the file.

ngxson commented:
>
> Sure, but we don't check between retries, there's a minor chance the
> content changes in between retries. It probably wasn't a big deal before,
> you pull a full file or you don't. But with resumable transfers it becomes
> more relevant, because you could have half one version of the file, half a
> different version of the file.


 we currently don't check between retries, but you can implement it. my
idea is that we can rely on ETag instead of Last-Modified.

the overall idea is: the first time the file is downloaded, ETag header is
pulled via the HEAD request and it should be stored somewhere. then, one of
3 cases may happen:
- file download is completed --> the next time user run the model, the
stored ETag is used in to verify if the file is up-to-date
- file download is failed --> check ETag in the next retry. I think this is
not yet implemented, so you can try adding this
- file download is half-completed --> when resume, the ETag is used to make
sure remote file content isn't changed (this case is not yet implemented
and should be added in the current PR)

ngxson commented on common/arg.cpp:
```diff
@@ -57,12 +57,18 @@ static std::string read_file(const std::string & fname) {
 }
 
 static void write_file(const std::string & fname, const std::string & content) {
```
if anything here fails, should we try to delete the `(file).tmp` ?

ericcurtin replied:
I guess... It was tempting to consolidate on fopen within this file too, but we can use both streams and fopen no big deal.

ngxson commented on common/arg.cpp:
```diff
@@ -217,250 +223,292 @@ struct curl_slist_ptr {
-        LOG_WRN("%s: HEAD invalid http status code received: %ld\n", __func__, http_code);
-        head_request_ok = false;
-    }
+        bool                               head_request_ok = false;
```
```suggestion
        bool head_request_ok = false;
```

ngxson commented on common/arg.cpp:
```diff
@@ -217,250 +223,292 @@ struct curl_slist_ptr {
-            LOG_WRN("%s: Last-Modified header is different (%s != %s): triggering a new download\n", __func__, last_modified.c_str(), headers.last_modified.c_str());
-            should_download = true;
+        // Initialize libcurl
+        curl_ptr       curl(curl_easy_init(), &curl_easy_cleanup);
```
```suggestion
        curl_ptr curl(curl_easy_init(), &curl_easy_cleanup);
```

ngxson commented on common/arg.cpp:
```diff
@@ -217,250 +223,292 @@ struct curl_slist_ptr {
-        if (remaining_attempts == 0) break;
-        std::this_thread::sleep_for(std::chrono::milliseconds(exponential_backoff_delay));
+// helper function to hide password in URL
+static std::string llama_download_hide_password_in_url(const std::string & url) {
```
can we simplify this function by using std::regex?

ericcurtin committed:
Add resumable downloads for llama-server model loading

- Implement resumable downloads in common_download_file_single function
- Add detection of partial download files (.downloadInProgress)
- Check server support for HTTP Range requests via Accept-Ranges header
- Implement HTTP Range request with "bytes=<start>-" header
- Open files in append mode when resuming vs create mode for new downloads

Signed-off-by: Eric Curtin <eric.curtin@docker.com>

ericcurtin commented:
A windows flake, that's rare:

```
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  7  239M    7 18.2M    0     0  33.1M      0  0:00:07 --:--:--  0:00:07 33.4M
 17  239M   17 41.5M    0     0  41.1M      0  0:00:05  0:00:01  0:00:04 41.3M
curl: (18) end of response with 207310552 bytes missing
No marker found, stopped after 1.00 MB.
```

ngxson reviewed (approved):
Seems to work on my side. Btw I think it would be nice if we can have a better progress display, as the progress counts back from 0 when I resume the download, even though it only downloads the missing part.

And btw since we already implemented `Range` here, we should also be able to implement multi-threaded downloading which should significantly improve the download speed. We should look into this in the future.

ericcurtin commented:
Btw @ngxson or @doringeman if you have a Windows machine around, I'd appreciate a quick test of this progress bar on Windows:

```
llama-run hf://ggml-org/gemma-3-4b-it-qat-GGUF
```

I hope to just port a version of that over to llama-server. I only tested Linux/macOS at the time.

ericcurtin commented:
> @ericcurtin Hi, sure. So it depends on the current codepage. For me the default was 437:
> 
> <img alt="Screenshot 2025-09-18 204348" width="1902" height="301" src="https://private-user-images.githubusercontent.com/131475237/491248456-1bb0e51f-24a6-4389-8873-14f3f2bbf817.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTgyMzAzMTQsIm5iZiI6MTc1ODIzMDAxNCwicGF0aCI6Ii8xMzE0NzUyMzcvNDkxMjQ4NDU2LTFiYjBlNTFmLTI0YTYtNDM4OS04ODczLTE0ZjNmMmJiZjgxNy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTE4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkxOFQyMTEzMzRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kMDY3ZmQ4ZmI0MmI0NDU3OTM5YzRhMDZjMTNjZTQ5NGUyZmExMmE4YTZjYTk4OGMyMGRkZTkyYjhhMTJkMGUwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.Kfm49n2is3Grn2M8laXPax5waL21pacPM9stEJ7EdSc">
> Maybe on Windows it is better to explicitly set it to UTF-8 like:
> 
> ```
> #ifdef _WIN32
>     UINT originalOutputCP = GetConsoleOutputCP();
>     SetConsoleOutputCP(CP_UTF8);
> #endif
> 
> ...
> 
> #ifdef _WIN32
>     SetConsoleOutputCP(originalOutputCP);
> #endif
> ```
> 
> Or use some ASCII character (like #) as a progressbar

Wanna open a PR? Since you are set up to test it on Windows. Number 1 looks pretties I think.

npopov-vst commented:
> Without the hashes looks prettier I think.

Agree.

> Wanna open a PR? Since you are set up to test it on Windows.

I can, but I am not sure where should I place these changes, since you mentioned in https://github.com/ggml-org/llama.cpp/pull/15988#issuecomment-3289519635, that you are going to refactor some parts.

I can confirm, that setting `SetConsoleOutputCP(CP_UTF8);` fixes the progress bar and I am happy to test your changes if needed.
Also It makes sense to add `SetConsoleCP(CP_UTF8);` in order to properly handle unicode input on Windows. So adding 
```
#ifdef _WIN32
    SetConsoleCP(CP_UTF8);
    SetConsoleOutputCP(CP_UTF8);
#endif
```
should solve all issues.
